### DISCUSSION TITLE
State change does not trigger View update

### DISCUSSION THREAD

**Discussion by ferologics** [pointfreeco/swift-composable-architecture#390] [Q&A] [answered]
I have a list with `ForEachStore` with nested state:
```swift
struct State {
  var messages: [MessageState]
}

struct MessageState {
  var state: State
}

extension MessageState {
  struct State {
    case `default`, loading, error
  }
}
```

When I update my `State.message[idx].state` from `loading` to `default` nothing happens. Only on top level `messages` change the view is updated to correct state (i.e. I append a new message).

What can I do to work around this?

Btw, it looks like [this issue on forums](https://forums.swift.org/t/swiftui-list-view-doesnt-get-updated-even-the-state-gets-updated/38171) is related.

**Comment by lukeredpath** [comment]
Can you post the corresponding views for `State` and `MessageState`?

**Comment by ferologics** [comment, marked as answer]
I tried to reproduce inside a [new project](https://github.com/pointfreeco/swift-composable-architecture/files/6031361/StateUpdateViewNot.zip) which led me to fixing the problem. The issue was that I used `.equatable()` on my own `WithViewStore` flavor called `WithEquatableViewStore`. This prevented my view from updating 😑

**Reply by lukeredpath** [reply]
I'm curious about why you need a version of `WithViewStore` that conforms to `Equatable`? Its already designed to only trigger a body re-calculation if the state changes when your state is `Equatable`. It only usually makes sense to implement `Equatable` on a view if its equability cannot be determined by the value of it's state alone.

**Reply by ferologics** [reply]
I was trying to squeeze any potential drops of performance out of the UI render pass cycle. 

SwiftUI potentially does some magic when it sees equatable views, though I have no confirmation of such behavior.

Yes you can remove duplicates in view store, but I wanted to test if there's any performance improvement in SwiftUI comparing against the EquatableView under the hood.

If there was any, it was not noticeable.
